File: fastify-ts-bun-backend/tests/request-context.test.ts
import type { FastifyInstance } from 'fastify';
import { afterEach, describe, expect, it } from 'vitest';
import { createApp } from '../src/app';

describe('request context', () => {
  let app: FastifyInstance | undefined;

  afterEach(async () => {
    if (app) {
      await app.close();
      app = undefined;
    }
  });

  it('propagates incoming x-request-id and emits trace headers', async () => {
    app = await createApp();
    const instance = app;
    const requestId = 'req-12345';

    const response = await instance.inject({
      method: 'GET',
      url: '/api/v1/ping',
      headers: {
        'x-request-id': requestId,
      },
    });

    expect(response.statusCode).toBe(200);
    expect(response.headers['x-request-id']).toBe(requestId);
    expect(response.headers['x-trace-id']).toBeDefined();
    expect(response.headers['x-span-id']).toBeDefined();
  });

  it('adds request and trace identifiers to error responses', async () => {
    app = await createApp();
    const instance = app;

    const response = await instance.inject({
      method: 'GET',
      url: '/route-does-not-exist',
    });

    expect(response.statusCode).toBe(404);
    const body = response.json();
    expect(body.requestId).toBeDefined();
    expect(body.traceId).toBeDefined();
  });
});

File: fastapi-python-backend/tests/request_context_test.py
from fastapi.testclient import TestClient


def test_propagates_incoming_request_id_and_emits_trace_headers(client: TestClient) -> None:
    request_id = 'req-12345'

    response = client.get('/api/v1/ping', headers={'x-request-id': request_id})

    assert response.status_code == 200
    assert response.headers['x-request-id'] == request_id
    assert response.headers['x-trace-id']
    assert response.headers['x-span-id']


def test_adds_request_and_trace_identifiers_to_error_responses(client: TestClient) -> None:
    response = client.get('/route-does-not-exist')

    assert response.status_code == 404
    assert response.json()['requestId']
    assert response.json()['traceId']
